File: src/lib/idol/functional.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.naiveObjUpdate = naiveObjUpdate;
exports.naiveObjectConcat = naiveObjectConcat;
exports.cachedProperty = cachedProperty;
exports.Alt = exports.StringSet = exports.OrderedObj = void 0;

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var OrderedObj =
/*#__PURE__*/
function () {
  function OrderedObj() {
    var obj = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    var ordering = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : Object.keys(obj).sort();

    _classCallCheck(this, OrderedObj);

    this.obj = obj;
    this.ordering = ordering;
  }

  _createClass(OrderedObj, [{
    key: "isEmpty",
    value: function isEmpty() {
      for (var _k in this.obj) {
        return false;
      }

      return true;
    }
  }, {
    key: "map",
    value: function map(f) {
      var _this = this;

      var newObj = {};
      this.ordering.forEach(function (k) {
        return newObj[k] = f(_this.obj[k]);
      });
      return new OrderedObj(newObj, this.ordering);
    }
  }, {
    key: "mapIntoIterable",
    value: function mapIntoIterable(f) {
      var _this2 = this;

      return this.ordering.map(function (k) {
        return f(k, _this2.obj[k]);
      });
    }
  }, {
    key: "mapAndFilter",
    value: function mapAndFilter(f) {
      var _this3 = this;

      return OrderedObj.fromIterable(this.keys().reduce(function (result, k) {
        return result.concat(f(_this3.obj[k]).map(function (v) {
          return [new OrderedObj(_defineProperty({}, k, v))];
        }).getOr([]));
      }, []));
    }
  }, {
    key: "concat",
    value: function concat(other) {
      var _this4 = this;

      var ordering = this.ordering.concat(other.ordering.filter(function (k) {
        return !(k in _this4.obj);
      }));
      var result = {};
      ordering.forEach(function (k) {
        var left = _this4.obj[k];
        var right = other.obj[k];
        if (!left) result[k] = right;else if (!right) result[k] = left;else result[k] = left.concat(right);
      });
      return new OrderedObj(result, ordering);
    }
  }, {
    key: "keys",
    value: function keys() {
      return Object.keys(this.obj);
    }
  }, {
    key: "concatMap",
    value: function concatMap(f, d) {
      var _this5 = this;

      return this.ordering.reduce(function (result, nextK) {
        return result.concat(f(nextK, _this5.obj[nextK]));
      }, d);
    }
  }, {
    key: "values",
    value: function values() {
      var _this6 = this;

      return this.ordering.map(function (k) {
        return _this6.obj[k];
      });
    }
  }, {
    key: "iter",
    value: function iter() {
      var _ref2;

      var i = 0;
      var obj = this.obj;
      var keys = this.keys();
      return _ref2 = {}, _defineProperty(_ref2, Symbol.iterator, function () {
        return this;
      }), _defineProperty(_ref2, "next", function next() {
        if (i < keys.length) {
          var key = keys[i];
          return {
            done: false,
            value: [key, obj[key]]
          };
        }

        return {
          done: true
        };
      }), _ref2;
    }
  }, {
    key: "get",
    value: function get(k) {
      if (k in this.obj) {
        return Alt.lift(this.obj[k]);
      }

      return Alt.empty();
    }
  }, {
    key: "length",
    get: function get() {
      return Object.keys(this.obj).length;
    }
  }], [{
    key: "fromIterable",
    value: function fromIterable(iter) {
      var result = new OrderedObj();
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = iter[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var o = _step.value;
          result = result.concat(o);
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator["return"] != null) {
            _iterator["return"]();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      return result;
    }
  }]);

  return OrderedObj;
}();

exports.OrderedObj = OrderedObj;

var StringSet =
/*#__PURE__*/
function () {
  function StringSet() {
    var _this7 = this;

    var items = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];

    _classCallCheck(this, StringSet);

    var obj = {};
    this.items = [];
    items.forEach(function (i) {
      if (i in obj) return;
      obj[i] = [];

      _this7.items.push(i);
    });
  }

  _createClass(StringSet, [{
    key: "concat",
    value: function concat(other) {
      return new StringSet(this.items.concat(other.items));
    }
  }]);

  return StringSet;
}();

exports.StringSet = StringSet;

var Alt =
/*#__PURE__*/
function () {
  function Alt(value) {
    _classCallCheck(this, Alt);

    this.value = value.slice(0, 1);
  }

  _createClass(Alt, [{
    key: "unwrap",
    value: function unwrap() {
      if (this.value.length) {
        return this.value[0];
      }

      throw new Error("Unwrapped empty value!");
    }
  }, {
    key: "getOr",
    value: function getOr(d) {
      if (this.value.length) {
        return this.value[0];
      }

      return d;
    }
  }, {
    key: "binding",
    value: function binding() {
      var value = this.value;
      return (
        /*#__PURE__*/
        regeneratorRuntime.mark(function _callee() {
          return regeneratorRuntime.wrap(function _callee$(_context) {
            while (1) {
              switch (_context.prev = _context.next) {
                case 0:
                  if (value.length) {
                    _context.next = 3;
                    break;
                  }

                  _context.next = 3;
                  return false;

                case 3:
                  return _context.abrupt("return", value[0]);

                case 4:
                case "end":
                  return _context.stop();
              }
            }
          }, _callee);
        })()
      );
    }
  }, {
    key: "bind",
    value: function bind(f) {
      if (this.isEmpty()) return this;
      return f(this.unwrap());
    }
  }, {
    key: "isEmpty",
    value: function isEmpty() {
      return !this.value.length;
    }
  }, {
    key: "concat",
    value: function concat(other) {
      if (this.isEmpty()) return other;
      return this;
    }
  }, {
    key: "map",
    value: function map(f) {
      if (this.isEmpty()) return this;
      return Alt.lift(f(this.unwrap()));
    }
  }, {
    key: "either",
    value: function either(other) {
      if (!this.isEmpty() && !other.isEmpty()) {
        throw new Error("Unexpected conflict, found ".concat(this.value.join(" ")));
      }

      return this.concat(other);
    }
  }, {
    key: "filter",
    value: function filter(pred) {
      if (this.isEmpty()) return this;
      if (pred(this.unwrap())) return this;
      return Alt.empty();
    }
  }], [{
    key: "from",
    value: function from(f) {
      var result;

      while (!(result = f.next()).done) {
        if (!result.value) {
          return new Alt([]);
        }
      }

      return new Alt([result.value]);
    }
  }, {
    key: "lift",
    value: function lift(v) {
      return new Alt([v]);
    }
  }, {
    key: "empty",
    value: function empty() {
      return new Alt([]);
    }
  }]);

  return Alt;
}();

exports.Alt = Alt;

function naiveObjUpdate(one, other) {
  for (var _k2 in other) {
    if (_k2 in one) one[_k2] = one[_k2].concat(other[_k2]);else one[_k2] = other[_k2];
  }
}

function naiveObjectConcat(one, other) {
  var result = _objectSpread({}, one);

  naiveObjUpdate(result, other);
  result.constructor = one.constructor;
  return result;
}

function cachedProperty(store, key, f) {
  key = "__" + key;
  if (key in store) return store[key];
  return store[key] = f();
}

File: src/lib/idol/functional.py
from typing import TypeVar, Generic, Tuple, List, Iterable, Any, Optional, Iterator, Mapping, Dict

A = TypeVar("A")
B = TypeVar("B")
C = TypeVar("C")

T = TypeVar("T")
R = TypeVar("R")


class StringSet(set):
    def add(self, element: str):
        super(StringSet, self).add(element)

    def __init__(self, items: Iterable[str] = None):
        super(StringSet, self).__init__(items or [])

    __add__ = set.union
    concat = __add__


class OrderedObj(Generic[T]):
    obj: Mapping[str, T]
    ordering: List[str]

    def __init__(self, obj: Mapping[str, T] = None, ordering: List[str] = None):
        self.obj = obj or {}
        self.ordering = ordering or sorted(self.obj.keys())

    @classmethod
    def from_iterable(cls, items: Iterable["OrderedObj[T]"]) -> "OrderedObj[T]":
        result = OrderedObj()
        for o in items:
            result += o
        return result

    def __bool__(self):
        return bool(self.obj)

    def __len__(self):
        return len(self.obj)

    def concat(self, other: "OrderedObj[T]") -> "OrderedObj[T]":
        ordering = self.ordering + [i for i in other.ordering if i not in self.obj]
        obj: Dict[str, T] = {}

        for k in ordering:
            if k not in self.obj:
                obj[k] = other.obj[k]
            elif k not in other.obj:
                obj[k] = self.obj[k]
            else:
                left = self.obj[k]
                right = other.obj[k]
                obj[k] = left + right

        return OrderedObj(obj, ordering)

    __add__ = concat

    def keys(self) -> Iterable[str]:
        return self.ordering

    def values(self) -> Iterable[T]:
        for k in self.ordering:
            yield self.obj[k]

    def __iter__(self) -> Iterable[Tuple[str, T]]:
        for k in self.ordering:
            yield k, self.obj[k]

    def get(self, k: str) -> "Alt[T]":
        if k in self.obj:
            return Alt.lift(self.obj[k])
        return Alt.empty()


class Conjunct(Generic[A]):
    v: List[A]

    def __init__(self, v: Iterable[A]):
        if not isinstance(v, list):
            v = list(v)
        self.v = v

    @classmethod
    def lift(cls, v: A) -> "Conjunct[A]":
        return cls((v,))

    @classmethod
    def empty(cls) -> "Conjunct[Any]":
        return cls([])

    def __iter__(self) -> Iterable[List[T]]:
        if self:
            yield self.v

    def __bool__(self):
        return len(self.v)

    def __concat__(self, other: "Conjunct[A]") -> "Conjunct[A]":
        return Conjunct(self.v + other.v)


class Alt(Generic[A]):
    v: List[A]

    def __init__(self, v: Iterable[A]):
        if not isinstance(v, list):

            def takeone():
                for i in v:
                    yield i
                    break

            v = list(takeone())
        self.v = v

    @classmethod
    def lift(cls, v: A) -> "Alt[A]":
        return cls((v,))

    @classmethod
    def lift_optional(cls, v: Optional[A]) -> "Alt[A]":
        if v is None:
            return cls.empty()
        return cls.lift(v)

    @classmethod
    def empty(cls) -> "Alt[Any]":
        return cls([])

    def unwrap(self) -> A:
        return self.v[0]

    def get_or(self, d: A) -> A:
        if not self:
            return d
        return self.unwrap()

    def get_or_fail(self, msg: str) -> A:
        if self:
            return self.unwrap()
        raise ValueError(msg)

    def __iter__(self) -> Iterable[A]:
        if self:
            yield self.unwrap()

    def concat(self, other: "Alt[A]") -> "Alt[A]":
        if self:
            return self
        return other

    __add__ = concat

    def __bool__(self):
        return len(self.v) > 0


class Disjoint(Generic[A]):
    v: List[A]

    def __init__(self, v: Iterable[A]):
        if not isinstance(v, list):
            v = list(v)
        self.v = v

    @classmethod
    def lift(cls, v: A) -> "Disjoint[A]":
        return cls((v,))

    @classmethod
    def lift_optional(cls, v: Optional[A]) -> "Disjoint[A]":
        if v is None:
            return cls.empty()
        return cls.lift(v)

    @classmethod
    def empty(cls) -> "Disjoint[Any]":
        return cls([])

    def get_or(self, d: A) -> A:
        if not self:
            return d
        return self.unwrap()

    def get_or_fail(self, msg: str) -> A:
        if self:
            return self.unwrap()
        raise ValueError(msg)

    def unwrap(self) -> A:
        if len(self.v) > 1:
            raise ValueError(f"Unexpected conflict found!")

        return self.v[0]

    def unwrap_errors(self) -> Iterator[List[A]]:
        if len(self.v) > 1:
            yield self.v

    def __iter__(self) -> Iterable[A]:
        if len(self.v) > 0:
            yield self.unwrap()

    def concat(self, other: "Disjoint[A]") -> "Disjoint[A]":
        return Disjoint(self.v + other.v)

    __add__ = concat

    def __bool__(self):
        return len(self.v) > 0


def naive_object_update(self, other: object) -> None:
    self_dict = self.__dict__
    for k, v in other.__dict__.items():
        if k in self_dict:
            self_dict[k] += v
        else:
            self_dict[k] = v


def naive_object_concat(self: A, other: A) -> A:
    result = self.__class__.__new__()
    result.__dict__ = dict(**self.__dict__)
    naive_object_update(result, other)
    return result
